### PR TITLE
Solana watcher

### LINKED ISSUE #7: Solana Support

### PR DESCRIPTION
**Pull request by heyitaki**
This PR adds a watcher for Solana. It uses the [`getSignaturesForAddress`](https://docs.solana.com/developing/clients/jsonrpc-api#getsignaturesforaddress) API to grab transactions that list the Wormhole core bridge contract under `accountKeys` and parses the associated instructions to retrieve the payload from the corresponding message account.

Highly recommended to use a faster RPC node than the default, even to run tests.

Resolves #7.

### CODE REVIEW

File: watcher/src/watchers/SolanaWatcher.ts [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+    const vaasByBlock: VaasByBlock = {};
+
+    // get transaction bounds, fromSignature occurs after toSignature because getSignaturesForAddress walks backwards
+    const toBlock = await this.connection.getBlock(toSlot, { maxSupportedTransactionVersion: 0 });
+    if (!toBlock || !toBlock.blockTime) throw new Error(`solana: failed to fetch block ${toSlot}`);
+    const fromSignature = toBlock.transactions.at(-1)?.transaction.signatures[0];
+
+    let fromBlock: VersionedBlockResponse | null = null;
+    try {
+      fromBlock = await this.connection.getBlock(fromSlot, { maxSupportedTransactionVersion: 0 });
```
Comment (ckeun) [inline watcher/src/watchers/SolanaWatcher.ts:48]: is there a risk of double evaluating blocks? e.g., start at (fromBlock=10, toBlock=11) <-- it looks like it's iterating through all of the transactions in both of these blocks... then does it go (fromBlock=11, toBlock=12) and re-evaluate the transactions in Block 11?

Comment (heyitaki) [reply watcher/src/watchers/SolanaWatcher.ts:48]: Within a single call of this method we should walk through every transaction once. You're right that this method is inclusive of `fromBlock` and `toBlock`, but the way `watch` works in `Watcher.ts`, `fromBlock` is incremented to one above `toBlock`, so we need to traverse all transactions in `fromBlock` in all cases but the first iteration when we fetch the last block from the db. We can (and should?) eliminate that case, by changing `fromBlock` to be the block after the last recorded block in the db.

Comment (ckeun) [reply watcher/src/watchers/SolanaWatcher.ts:48]: ah okay, so you're going 2x2.  👍

File: watcher/src/watchers/SolanaWatcher.ts [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+        }
+      }
+
+      numSignatures = signatures.length;
+      currSignature = signatures.at(-1)?.signature;
+    }
+
+    // add last block for storeVaasByBlock
+    const lastBlockKey = makeBlockKey(
+      toSlot.toString(),
```
Comment (ckeun) [inline watcher/src/watchers/SolanaWatcher.ts:122]: should the lastBlockKey be using fromSlot.toString()? in case it crashes, it'll make sure to re-evaluate the fromSlot() again?

File: watcher/src/watchers/SolanaWatcher.ts [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+      new Date(toBlock.blockTime * 1000).toISOString()
+    );
+    return Object.assign({ [lastBlockKey]: [] }, vaasByBlock);
```
Comment (evan-gray) [inline watcher/src/watchers/SolanaWatcher.ts:125]: Let's just return `{ [lastBlockKey]: [], ...vaasByBlock}`?

Comment (heyitaki) [inline watcher/src/watchers/SolanaWatcher.ts:125]: We should either reverse keys before returning (since they are currently returned in reverse chronological order) or search `vaasByBlock` for the last block in `storeVaasByBlock`. After discussion, the second approach is preferred because it is more modular.

File: watcher/src/watchers/__tests__/SolanaWatcher.test.ts [hunk 4]
```diff
@@ -0,0 +1,74 @@
+import dotenv from 'dotenv';
```
Comment (heyitaki) [inline watcher/src/watchers/__tests__/SolanaWatcher.test.ts:1]: Would be nice to add a test to ensure that repeated calls of `getMessagesForBlocks` on adjacent block ranges don't duplicate (or drop) messages

Comment (heyitaki) [reply watcher/src/watchers/__tests__/SolanaWatcher.test.ts:1]: Also a test to validate `blockKey` and `vaaKey` (to ensure things like human readable timestamp instead of unix timestamp 😛)

File: watcher/src/watchers/SolanaWatcher.ts (hunk 3 above)
Comment (heyitaki) [reply watcher/src/watchers/SolanaWatcher.ts]: Fixed

Comment (heyitaki) [reply watcher/src/watchers/SolanaWatcher.ts]: Fixed

File: watcher/src/watchers/__tests__/SolanaWatcher.test.ts (hunk 4 above)
Comment (heyitaki) [reply watcher/src/watchers/__tests__/SolanaWatcher.test.ts:1]: Both added

**Review by ckeun — APPROVED** [review]

File: watcher/src/utils/aptos.ts [hunk 5]
```diff
@@ -0,0 +1,30 @@
+import { INITIAL_DEPLOYMENT_BLOCK_BY_CHAIN } from '@wormhole-foundation/wormhole-monitor-common';
+import { z } from 'zod';
+
+export const isValidBlockKey = (key: string) => {
```
Comment (evan-gray) [inline watcher/src/utils/aptos.ts:4]: I think these could be implemented as static methods on the corresponding classes, that way the aptos one can override the default one (which seems like what you're going for).

File: watcher/src/watchers/Watcher.ts [hunk 6]
```diff
@@ -25,7 +25,8 @@ export class Watcher {
 
   async watch(): Promise<void> {
     let toBlock: number | null = null;
-    let fromBlock: number | null = await getLastBlockByChain(this.chain);
+    const lastBlock = await getLastBlockByChain(this.chain);
+    let fromBlock: number | null = lastBlock ? lastBlock + 1 : null;
```
Comment (evan-gray) [inline watcher/src/watchers/Watcher.ts:29]: This makes me feel better, but also sad to have messed it up.

File: watcher/src/watchers/SolanaWatcher.ts [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+export class SolanaWatcher extends Watcher {
+  connection: Connection;
+  getSignaturesLimit = GET_SIGNATURES_LIMIT; // modifiable in tests
+
+  constructor() {
+    super('solana');
+    this.connection = new Connection(RPCS_BY_CHAIN.solana!, COMMITMENT);
+
+    // can be arbitrarily large since getMessagesForBlocks handles pagination internally
+    this.maximumBatchSize = 100000;
```
[the remark this answers is not in the document]
Comment (evan-gray) [reply watcher/src/watchers/SolanaWatcher.ts:29]: @panoel would like underscores for "older-eyes"

File: watcher/src/watchers/SolanaWatcher.ts [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+        {
+          before: currSignature,
+          until: toSignature,
+          limit: this.getSignaturesLimit,
+        }
+      );
+
+      this.logger.info(`processing ${signatures.length} transactions`);
+      for (const { signature } of signatures) {
+        const res = await this.connection.getTransaction(signature, {
```
Comment (evan-gray) [inline watcher/src/watchers/SolanaWatcher.ts:75]: Can you add a comment at the top of this block that explains the process here.

File: watcher/src/watchers/SolanaWatcher.ts [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+            this.chain,
+            emitterAddress.toString('hex'),
+            sequence.toString()
+          );
+          vaasByBlock[blockKey] = [...(vaasByBlock[blockKey] || []), vaaKey];
+        }
+      }
+
+      numSignatures = signatures.length;
+      currSignature = signatures.at(-1)?.signature;
```
Comment (evan-gray) [inline watcher/src/watchers/SolanaWatcher.ts:119]: Can we double check that this should be the last one?

File: watcher/src/watchers/__tests__/SolanaWatcher.test.ts [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+dotenv.config();
+
+import { describe, expect, jest, test } from '@jest/globals';
+import { SolanaJSONRPCError } from '@solana/web3.js';
+import { INITIAL_DEPLOYMENT_BLOCK_BY_CHAIN } from '@wormhole-foundation/wormhole-monitor-common/dist/consts';
+import { isValidVaaKey } from '../../utils/solana';
+import { isValidBlockKey } from '../../utils/validation';
+import { SolanaWatcher } from '../SolanaWatcher';
+
+jest.setTimeout(6000000);
```
Comment (evan-gray) [inline watcher/src/watchers/__tests__/SolanaWatcher.test.ts:11]: umm...

File: watcher/src/databases/BigtableDatabase.ts [hunk 11]
```diff
@@ -117,7 +117,7 @@ export class BigtableDatabase extends Database {
 
     if (updateLatestBlock) {
       // store latest vaasByBlock to firestore
-      const blockInfos = Object.keys(vaasByBlock);
+      const blockInfos = Object.keys(vaasByBlock).sort();
```
[the remark this answers is not in the document]
Comment (heyitaki) [reply watcher/src/databases/BigtableDatabase.ts]: Fixed, we now parse the block number and use that to compare

File: watcher/src/databases/JsonDatabase.ts [hunk 12]
```diff
@@ -50,10 +50,11 @@ export class JsonDatabase extends Database {
       this.db[chainId] = { ...(this.db[chainId] || {}), ...filteredVaasByBlock };
       writeFileSync(this.dbFile, JSON.stringify(this.db), ENCODING);
     }
+
     // this will always overwrite the "last" block, so take caution if manually backfilling gaps
-    const blockInfos = Object.keys(vaasByBlock);
+    const blockInfos = Object.keys(vaasByBlock).sort();
```
[the remark this answers is not in the document]
Comment (heyitaki) [reply watcher/src/databases/JsonDatabase.ts]: Fixed

File: watcher/src/types/aptos.ts [hunk 13]
```diff
@@ -0,0 +1,13 @@
+import { Types } from 'aptos';
+
+export type AptosEvent = Omit<Types.Event, 'data'> & {
```
[the remark this answers is not in the document]
Comment (heyitaki) [reply watcher/src/types/aptos.ts:3]: Where did you come from cotton eye joe

File: watcher/src/utils/aptos.ts (hunk 5 above)
Comment (heyitaki) [reply watcher/src/utils/aptos.ts]: Moved to watchers, although decided not to make them static as the base `isValidBlockKey` uses the `chain` field. Can make watcher pass in chain as a param and make these methods static if you have a strong opinion

File: watcher/src/watchers/SolanaWatcher.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (heyitaki) [reply watcher/src/watchers/SolanaWatcher.ts]: Fixed

File: watcher/src/watchers/SolanaWatcher.ts [hunk 14]
```diff
@@ -0,0 +1,129 @@
...
+import { isLegacyMessage, normalizeCompileInstruction } from '../utils/solana';
+import { Watcher } from './Watcher';
+
+const WORMHOLE_PROGRAM_ID = CONTRACTS.MAINNET.solana.core;
+const COMMITMENT: Commitment = 'finalized';
+const GET_SIGNATURES_LIMIT = 1000;
+
+export class SolanaWatcher extends Watcher {
+  connection: Connection;
+  getSignaturesLimit = GET_SIGNATURES_LIMIT; // modifiable in tests
```
[the remark this answers is not in the document]
Comment (heyitaki) [reply watcher/src/watchers/SolanaWatcher.ts]: Added

File: watcher/src/watchers/__tests__/SolanaWatcher.test.ts (hunk 10 above)
Comment (heyitaki) [reply watcher/src/watchers/__tests__/SolanaWatcher.test.ts]: Fixed 😅

File: watcher/src/watchers/Watcher.ts (hunk 6 above)
Comment (evan-gray) [reply watcher/src/watchers/Watcher.ts:53]: Wait a minute... this is now wrong on the initial block.

**Review by evan-gray — CHANGES_REQUESTED** [review]

**Comment by heyitaki** [comment]
@egray you're right it should be 
```
let fromBlock: number = lastBlock ? lastBlock + 1 : INITIAL_DEPLOYMENT_BLOCK_BY_CHAIN.solana;
```

File: common/src/consts.ts [hunk 15]
```diff
@@ -22,4 +22,5 @@ export const INITIAL_DEPLOYMENT_BLOCK_BY_CHAIN: {
   aptos: '0', // block is 1094390 but AptosWatcher uses sequence number instead
   near: '72767136',
   xpla: '777549',
+  solana: '96383501',
```
[the remark this answers is not in the document]
Comment (heyitaki) [reply common/src/consts.ts:25]: Hm I wrote a script to fetch it, perhaps it was only looking for txs with `postMessage` instructions...

File: watcher/src/watchers/SolanaWatcher.ts [hunk 16]
```diff
@@ -0,0 +1,157 @@
...
+  }
+
+  async getMessagesForBlocks(fromSlot: number, toSlot: number): Promise<VaasByBlock> {
+    if (fromSlot > toSlot) throw new Error('solana: invalid block range');
+    this.logger.info(`fetching info for blocks ${fromSlot} to ${toSlot}`);
+    const vaasByBlock: VaasByBlock = {};
+
+    // identify block range by fetching the first and last transactions
+    // getSignaturesForAddress walks backwards so fromSignature occurs after toSignature
+    const toBlock = await this.connection.getBlock(toSlot, { maxSupportedTransactionVersion: 0 });
```
Comment (evan-gray) [inline watcher/src/watchers/SolanaWatcher.ts:49]: Not having this in a try/catch like the `fromBlock` results in an error on backfilling when it naturally lands on a skipped `toBlock`

**Review by evan-gray — CHANGES_REQUESTED** [review]

**Comment by evan-gray** [comment]
Fixed all outstanding issues.

**Review by evan-gray — APPROVED** [review]
